### PR TITLE
Deprecate ffdhe

### PR DESCRIPTION
**Pull request by nimia**

### CODE REVIEW

File: draft-ietf-tls-deprecate-obsolete-kex.md [hunk 1]
```diff
@@ -215,26 +206,27 @@ of implementing security countermeasures correctly.
...
-for negotiating the group size, and some implementations only support small
-group sizes (see {{!RFC7919}}, Section 1).
+1. FFDHE suffers from interoperability problems because there is no mechanism for
+negotiating the group, and some implementations only support small group sizes
+(see {{!RFC7919}}, Section 1).
 
-2. In practice, some operators use 1024-bit FFDHE groups since this is the
+2. FFDHE groups may have small subgroups, which enables several attacks
+{{subgroups}}. When presented with a custom, non-standardized FFDHE group, a handshaking client cannot practically verify that the group chosen by the server does not suffer from this problem. There is also no mechanism for such handshakes to fall back to other key exchange parameters that are acceptable to the client.
+Custom FFDHE groups are widespread (as a result of advice based on {{weak-dh}}).
```
Comment (carrickdb) [inline draft-ietf-tls-deprecate-obsolete-kex.md]: Maybe explicitly state why this is bad? Something like "Custom FFDHE groups are widespread (as a result of advice based on {{weak-dh}}), which means that clients cannot simply reject handshakes that present custom and thus potentially dangerous groups."

File: draft-ietf-tls-deprecate-obsolete-kex.md [hunk 2]
```diff
@@ -259,40 +251,21 @@ Given these problems, this document updates {{!RFC4346}}, {{!RFC5246}},
 
 # Non-Ephemeral Diffie Hellman {#non-ephemeral}
 
-Clients MUST NOT offer non-ephemeral FFDH cipher suites in TLS 1.2 connections.
+Clients and servers MUST NOT offer non-ephemeral FFDH cipher suites in TLS 1.2 connections.
```
Comment (carrickdb) [inline draft-ietf-tls-deprecate-obsolete-kex.md]: Nit: If I'm not mistaken, the server doesn't offer cipher suites, per se, but selects one. (From the specification of the ServerHello in RFC 5246: "The single cipher suite selected by the server from the list in ClientHello.cipher_suites.") Thus, I suggest: "Clients MUST NOT offer and servers MUST NOT select non-ephemeral..."

File: draft-ietf-tls-deprecate-obsolete-kex.md [hunk 3]
```diff
@@ -259,40 +251,21 @@ Given these problems, this document updates {{!RFC4346}}, {{!RFC5246}},
...
 (Note that TLS 1.0 and 1.1 are deprecated by {{!RFC8996}} and TLS 1.3 does not
 support FFDH {{!RFC8446}}.) This includes all cipher suites listed in the table in
 {{appendix-dh}}.
 
-Clients SHOULD NOT offer non-ephemeral ECDH cipher suites in TLS 1.2
+Clients and servers SHOULD NOT offer non-ephemeral ECDH cipher suites in TLS 1.2
 connections. (Note that TLS 1.0 and 1.1 are deprecated by {{!RFC8996}} and
```
Comment (carrickdb) [inline draft-ietf-tls-deprecate-obsolete-kex.md]: Similar comment here as above.

File: draft-ietf-tls-deprecate-obsolete-kex.md [hunk 4]
```diff
@@ -259,40 +251,21 @@ Given these problems, this document updates {{!RFC4346}}, {{!RFC5246}},
...
-
-We note that, previously, supporting the broadest range of clients would have
-required supporting either RSA key exchange or 1024-bit FFDHE. This is no longer
-the case, and it is possible to support most clients released since circa 2015
-using 2048-bit FFDHE or more modern key exchange methods, and without RSA key
-exchange {{server_side_tls}}.
-
-All the cipher suites that do not meet the above requirements are listed in the
-table in {{appendix-dhe}}.
+Clients and servers MUST NOT offer FFDHE cipher suites in TLS 1.2 connections.
```
Comment (carrickdb) [inline draft-ietf-tls-deprecate-obsolete-kex.md]: Same comment here as above.

File: draft-ietf-tls-deprecate-obsolete-kex.md [hunk 5]
```diff
@@ -259,40 +251,21 @@ Given these problems, this document updates {{!RFC4346}}, {{!RFC5246}},
...
+This includes all cipher suites listed in the table in {{appendix-dhe}}.
+(Note that TLS 1.0 and 1.1 are deprecated by {{!RFC8996}}.) FFDHE cipher suites in TLS 1.3 do not suffer from the above problems {{!RFC8446}}. Therefore, clients and servers MAY offer FFDHE cipher suites in TLS 1.3 connections
```
Comment (carrickdb) [inline draft-ietf-tls-deprecate-obsolete-kex.md]: Nit: There are a lot of problems mentioned above, so maybe link to the section that mentions these particular problems?

Comment (carrickdb) [inline draft-ietf-tls-deprecate-obsolete-kex.md]: Nit: There seems to be a missing period at the end of this sentence.

File: draft-ietf-tls-deprecate-obsolete-kex.md [hunk 6]
```diff
@@ -464,7 +440,7 @@ comments and suggestions.
 | TLS_ECDH_RSA_WITH_CAMELLIA_128_GCM_SHA256 | {{!RFC6367}} |
 | TLS_ECDH_RSA_WITH_CAMELLIA_256_GCM_SHA384 | {{!RFC6367}} |
 
-# DHE Cipher Suites Refered to by This Document {#appendix-dhe}
+# DHE Cipher Suites deprecated by This Document {#appendix-dhe}
```
Comment (carrickdb) [inline draft-ietf-tls-deprecate-obsolete-kex.md:443]: 🎉

File: draft-ietf-tls-deprecate-obsolete-kex.md [hunk 7]
```diff
@@ -343,6 +317,8 @@ key reuse not only prevents decryption in the event of key compromise, but also
 precludes this category of attacks altogether. Therefore, this document
 discourages the reuse of elliptic curve DH public keys.
 
+As for ephemeral finite field Diffie-Hellman in TLS 1.2, as explained above, clients have no practical way to support these cipher suites while ensuring they only negotiate security parameters that are acceptable to them. In TLS 1.2, the server chooses the Diffie-Hellman group, and custom groups are prevalent. Therefore, once the client includes these cipher suites in its handshake and the server presents a custom group, the client cannot complete the handshake while ensuring security. Verifying the group structure is prohibitively expensive for the client. Using a safelist of known-good groups is also impractical, since server operators were encouraged to generate their own custom group. Further, there is no mechanism for the handshake to fall back to other parameters, that are acceptable to both the client and server.
```
Comment (carrickdb) [inline draft-ietf-tls-deprecate-obsolete-kex.md:320]: Nice summary of the issues!

**Review by carrickdb — APPROVED** [review]

File: draft-ietf-tls-deprecate-obsolete-kex.md (hunk 1 above)
Comment (nimia) [reply draft-ietf-tls-deprecate-obsolete-kex.md]: Done, thanks :-)

File: draft-ietf-tls-deprecate-obsolete-kex.md (hunk 2 above)
Comment (nimia) [reply draft-ietf-tls-deprecate-obsolete-kex.md]: Done, thanks :-)

File: draft-ietf-tls-deprecate-obsolete-kex.md (hunk 3 above)
Comment (nimia) [reply draft-ietf-tls-deprecate-obsolete-kex.md]: Done, thanks :-)

File: draft-ietf-tls-deprecate-obsolete-kex.md (hunk 5 above)
Comment (nimia) [reply draft-ietf-tls-deprecate-obsolete-kex.md]: I hope I added this correctly, I can't generate the HTML version for a few days... Can you run make successfully on your repo?

Comment (nimia) [reply draft-ietf-tls-deprecate-obsolete-kex.md]: Done, thanks :-)

File: draft-ietf-tls-deprecate-obsolete-kex.md (hunk 6 above)
Comment (nimia) [reply draft-ietf-tls-deprecate-obsolete-kex.md:443]: haha indeed!

File: draft-ietf-tls-deprecate-obsolete-kex.md (hunk 7 above)
Comment (nimia) [reply draft-ietf-tls-deprecate-obsolete-kex.md:320]: Thanks! :-)

**Pull request merged**